### PR TITLE
Upgrade to Electron 4

### PR DESCRIPTION
**Pull request by nathansobo**
This PR will track our progress upgrading to Electron 4.

We just upgraded to Electron 3, but we're experiencing a crash deep in the bowels of V8 (https://github.com/atom/atom/issues/19372). I'm interested in whether Electron 4 might fix this crash.

## Issues to resolve

- [x] https://github.com/atom/atom/issues/19374 Crash in mksnapshot when building with Electron 4
- [x] Update any native modules Atom is using to make sure `win_delay_load_hook` isn't false: https://electronjs.org/docs/tutorial/using-native-node-modules#a-note-about-win_delay_load_hook
  - [x] @atom/nsfw: atom/nsfw#6
  - [x] git-utils: atom/git-utils#94
  - [x] scandal (git-utils version bump): atom/scandal#47
- [x] Failing integration test
- [x] Native tabs super slow (https://github.com/atom/atom/issues/19611).

### CODE REVIEW

File: package.json
Comment (mfonville) [inline package.json]: Isn't it more practical to use 4.2.2 to also have the latest bugfixes?

**Comment by 50Wliu** [comment]
I was able to successfully build Atom by skipping the snapshot generation and using a patched version of fuzzy-native (to bypass node-pre-gyp), but starting Atom fails with "The specified procedure cannot be found" when attempting to load nsfw 😕. That usually points to a rebuild problem but apm seems to be doing the right thing...

File: package.json
Comment (UziTech) [reply package.json]: Why not go straight to `v5.0.1`?

Comment (nathansobo) [reply package.json]: It's just about being incremental. We tried to go to 3 first, but we're hitting https://github.com/atom/atom/issues/19372, so we wanted to see how we'd do in 4. Unfortunately we still don't really know.

Comment (nathansobo) [reply package.json]: But yeah, we should be on the latest patch release.

**Comment by Asday** [comment]
Just to make sure, this finally fixes #15323, right?

File: package.json
Comment (mfonville) [reply package.json]: 4.2.3 has been released yesterday

Comment (mfonville) [reply package.json]: And 4.2.4 has been tagged

**Comment by rafeca** [comment]
I fixed the remaining build failure on the smoke test, so now we got a green build for Electron v4!! 🎉

@50Wliu: are we able to merge https://github.com/atom/git-utils/pull/94 and publish a new "normal" version of `git-utils` to use it already on `master` or will this break on Electron v3? If it won't, it would be good to get these done.

I can help with that if you don't have bandwidth at the moment.

**Comment by 50Wliu** [comment]
@rafeca Great work!

Yeah, atom/git-utils#94 can be merged and used on master. Same with scandal. Would be appreciated if you could help me with that - I won't be able to get to it until tonight at the earliest.

**Comment by rafeca** [comment]
Thanks for the quick response @50Wliu! I'm going to do the releases then 😄

File: package.json
Comment (mfonville) [reply package.json]: And 4.2.5 has been tagged now too

**Comment by rafeca** [comment]
## General testing

- [ ] Review issues tagged with [`electron`](https://github.com/search?utf8=%E2%9C%93&q=org%3Aatom+label%3Aelectron+is%3Aopen+is%3Aissue) to see if any of them are fixed
- [ ] Review any comment in the Atom code that contains the keyword `TodoElectronIssue` to see if any of them can be resolved ([search link](https://github.com/search?q=org%3Aatom+TodoElectronIssue&type=Code)).
- [ ] Verify that you can collaborate with Teletype
	- [ ] Host using Atom's current stable release can collaborate with guest using Atom with the new version of Electron
	- [ ] Guest using Atom's current stable release can collaborate with host using Atom with the new version of Electron
	- [ ] Host and guest can collaborate when both are using Atom with the new version of Electron
  <details>
  <summary>Steps</summary>
	1. Open a file
	1. Share a portal
	1. Have a guest join the portal
	1. When guest edits the text in the file, verify that the host sees those changes
	</details>

## Check for regressions experienced in previous upgrades

### Text Input/Keybindings

See how to [setup keyboard layouts](https://github.com/atom/maintainers-log/blob/master/how-tos/setup-keyboard-layouts.md).

- [ ] IME not working (https://github.com/atom/atom/issues/14911)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Pinyin simplified  
  1. Open Atom and use the layout you installed as the keyboard layout
  1. Example, if you picked Pinyin simplified type in `zhongwenshuru`

  **Expected:** in Pinyin simplified `中文输入` is expected

  gif

  **Issues we've seen:** Only shows latin first character and not every character typed.

  gif
  </details>

- [ ] macOS: Key binding with composition characters are ignored (https://github.com/atom/atom/issues/15189)
  <details>
  <summary>Repro steps</summary>

  1. Open your keymap.cson file and add the following:
      ```cson
      'atom-text-editor':
        # use hjkl; vim-like when pressing alt (word by word, paragraph by paragraph)
        # or alt (letter by letter)
        'alt-h': 'core:move-left'
        'alt-j': 'core:move-down'
        'alt-k': 'core:move-up'
        'alt-l': 'core:move-right'
      ```
  1. Save keymap.cson and reload atom.
  1. Try the new keymappings under `ABC - Extended` keyboard layout

  **Expected**: the keybindings work

  **Issues we've seen:** it types out the modified keys and ignores the key mapping
  </details>

- [ ] macOS: Composition characters mess up insertion point (https://github.com/atom/atom/issues/15344)
  <details>
  <summary>Repro steps</summary>

  1. Open keymap.cson and add the following:
      ```cson
      'atom-text-editor':
        'alt-n': 'core:move-down'
        'alt-o': 'core:move-down'
      ```
  1. Have the following in your buffer:
      ```
      Foo
      Bar
      ```
  1. Have the cursor at the beginning of file(before F character) Type alt-n, then arrow right, then insert a character. Observe that the character is inserted at the beginning of the line (where the alt-n left you).
  1. Repeat steps using <kbd>alt-o</kbd> instead of <kbd>alt-n</kbd>.

  **Expected**: Both alt-o and alt-n behave the same

  **Issues we've seen:**
  bug

- [ ] Ubuntu: Keystrokes involving <kbd>ctrl</kbd> resolve to the default layout instead of the active layout (https://github.com/atom/atom/issues/13170)

  <details>
  <summary>Repro steps</summary>

  **This should be tested on Linux with gnome**

  1. Install a Linux ditro with gnome, for example Ubuntu
  1. Install French AZERTY and English US QWERTY layouts
  1. Change the default layout to be French AZERTY
  1. Open Atom and switch to English US QWERTY layout
  1. Press ctrl-w

  **Expected**: `core:close` to dispatch (Or the keybinding-resolver to resolve to ctrl-w). Should be resolving to the keyboard layout that is chosen and not the OS default layout.

  **Issues we've seen:** `core:undo` dispatched because it was resolved as ctrl-z because AZERTY has Z where W is on QWERTY.
  </details>

- [ ] Other keyboard layouts on new Electron version
  - [ ] macOS: ABC - Extended
  - [ ] macOS: Slovak
  - [ ] macOS: U.S. International
  - [ ] macOS: Squirrel (https://github.com/rime/squirrel)
  - [ ] macOS: Full-Width Alphanumeric (Google) - Japanese layout  (https://www.google.co.jp/ime/)
  - [ ] Windows: U.S.
  - [ ] Linux: Pinyin Simplified

### UI

- [ ] tree-view drag image (https://github.com/atom/tree-view/pull/1054)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a project in Atom with multiple files and a few folders
  1. Drag a file into a folder

  **Expected:**  
  image

  **Issues we've seen:**

  image
  </details>

- [ ] drag-and-drop indicator (https://github.com/atom/tree-view/pull/1055, https://github.com/atom/tabs/pull/426, https://github.com/atom/tabs/pull/437)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a few folder projects on tree-view (you can do this on terminal in a directory that has existing folder and having a space between each folder. Example: `atom folderA folderB folderC`)
  1. Drag and drop all of the folders to reorder them on Atom

  **Expected:**
  The placeholder indicates where the folder/project is being dropped to.

  screen shot 2017-03-29 at 16 31 16

  **Issues we've seen:**
  No placeholder shows up after drag and dropping

  screen shot 2017-03-29 at 16 30 51
  </details>

- [ ] Large file rendering (https://github.com/atom/atom/issues/16591)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has over one million lines
  1. Scroll down to the bottom

  **Expected:** To not regress the number of lines that can be rendered. Rendering to be correct for lines past a certain point. Atom 1.25  can render around 800k to 1 million lines correctly.

  **Issues we've seen:** Increased number of lines rendered but bad rendering past a certain point

  bad renderer
  </details>

- [ ] Loss of subpixel AA when the cursor is at the end of long lines (https://github.com/atom/atom/issues/16889, https://github.com/atom/atom/pull/16595)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open [text-editor.js](https://github.com/atom/atom/blob/2e19cf4a7c0a4d624b1f9c1072cd292de112cdb0/src/text-editor.js) in Atom
  1. Copy the first 21 lines to a new untitled window
  1. Turn on soft wrapping and change language mode to javascript
  1. Resize the Window so it's soft wrapped
  1. On line 15 type `a` until the you reach the end of the window

  **Expected:** To not lose subpixel AA

  **Issues we've seen:** Loss of subpixel AA. Both when soft wrapping was [enabled](https://github.com/atom/atom/pull/16943) and [disabled](https://github.com/atom/atom/pull/16595).

  subpixel aa loss softwrap
  </details>

- [ ] Scrolling horizontally shift + scroll wheel (https://github.com/atom/atom/pull/12696#issuecomment-288615912)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open a file that has long lines
  1. Turn off soft wrapping and resize the window so you have a horizontal scrollbar
  1. Press shift and scroll in both directions using the mouse wheel

  **Expected:** File to scroll horizontally in both directions

  **Issues we've seen:** File does not scroll horizontally at all in any direction
  </details>

- [ ] Scrollbars misbehaving on the first file that is opened (https://github.com/atom/atom/pull/12696#issuecomment-251299649)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom
  1. Open some file that has content enough to have a vertical scrollbar
  1. Close Atom
  1. Open Atom and look at the vertical scrollbar

  **Expected:** Scrollbar to be visible

  **Issues we've seen:** Scrollbar is not visible and is flickering when you are editing
  </details>

- [ ] Middle clicking on unsaved tab (https://github.com/atom/atom/issues/15197)
	- [ ] macOS
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Open Atom and open a new unsaved file.
  1. Make edits to the unsaved file and middle click on the tab.

  **Expected:**
  Clicking save/cancel or the options on the dialog works.

  **Issues we've seen:**
  The UI and the dialog is unable to receive mouse clicks. You can still choose options via Keyboard, but not mouse.
  </details>

- [ ] Linux: Atom scrolls even when not focused (https://github.com/atom/atom/issues/15482)
  <details>
  <summary>Repro steps</summary>

  1. Open Atom with a large enough file that has a scrollbar
  1. Open a different application and place it over the Atom window
  1. Scroll with the mouse inside the other window
  1. Focus the Atom window

  **Expected:** Atom window should keep the original scroll position

  **Issues we've seen:** The Atom window scrolls after it is focused
  </details>

### Other

- [ ] Deprecation warnings (https://github.com/atom/atom/pull/12696#issuecomment-288871751)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  **This should be tested on macOS, Linux and Windows with community packages installed**

  1. Open Atom
  1. Open the Developer tools console on the master branch and check for deprecation warnings
  1. Open the Developer tools console on the electron upgrade branch and check for deprecation warnings
  1. Compare the two outputs

  **Expected:** No new deprecation warnings

  **Issues we've seen:** New deprecation warnings both from core and community packages
  </details>

- [ ] Supported Versions of OS (https://github.com/atom/atom/issues/15297)
	- [ ] macOS
	- [ ] Windows 10
	- [ ] Ubuntu
  <details>
  <summary>Repro steps</summary>

  1. Check the list of supported platforms [Electron Master](https://github.com/electron/electron/blob/master/docs/tutorial/support.md#supported-platforms)
  1. Check what version is listed as minimum for each platform on https://atom.io and https://atom.io/faq

  **Expected:** https://atom.io and https://atom.io/faq to display the same versions as the Electron documentation

  **Issues we've seen:** macOS version support changed and https://atom.io and https://atom.io/faq was not updated
  </details>

- [ ] macOS : Native tabs (https://github.com/atom/atom/pull/14711)
  <details>
  <summary>Repro steps</summary>

  1. Open two atom windows on mac
  1. Go to View in menus and check `Show Tab Bar`
  screen shot 2017-06-03 at 5 10 32 pm
  1. Do this for both Atom windows.
  1. Drag one atom window to the other via tab bar

  **Expected:**
  Native Tab bar shows up and able to drag different windows to each other.  Also make sure UI isn't messed up.

  **Issues we've seen:**
  Native Tab Bar doesn't work and messes up UI when enabled
  </details>

- [ ] Ubuntu with KDE: Menu uses 100% CPU (https://github.com/electron/electron/issues/8455, https://github.com/atom/atom/issues/13885)
  <details>
  <summary>Repro steps</summary>

  1. Install Atom on Ubuntu KDE. [[instructions](https://linuxconfig.org/how-to-install-kde-plasma-desktop-on-ubuntu-18-04-bionic-beaver-linux)]
  1. Enable KDE plasma global menu: Right-click on the desktop -> Add Panel => Application Menu Bar. [[demo](https://youtu.be/qo5fhiXAUbI)]
  1. Start Atom.
  1. Try to use the desktop menu.

  **Expected:**
  Desktop menu items work

  **Issues we've seen:**
  Desktop menu items do nothing when you click on things.
  </details>

- [ ] macOS: Slovak QWERTZ (https://github.com/atom/atom-keymap/issues/223)
  <details>
  <summary>Repro steps</summary>

  1. Install Slovak keyboard layout on mac
  1. Open keyboard viewer to see the keys are are suppose to show up and open keyboard resolver on atom.
  1. hit `cmd+'`.

  **Expected:**  Resolves to `cmd+'` like it shows on keyboard viewer on mac.

  **Issues we've seen:** It resolves to `ctrl+§`
  </details>

- [ ] macOS: IME jump (https://github.com/atom/atom/issues/15696)
  <details>
  <summary>Repro steps</summary>

  1. Install an IME layout, for example Chinese pinyin
  1. Type one character

  **Expected:**
  Expected the IME to be positioned at under the character you entered
  **Issues we've seen:** The IME window is in the top left corner. When you enter the second character it jumps to be positioned under the text
  </details>

**Comment by jasonrudolph** [comment]
@rafeca: Since [Electron 4 drops support for macOS 10.9](https://github.com/electron/electron/blob/4-2-x/docs/tutorial/support.md), is there something we can do to ensure that Atom users on macOS 10.9 don't get auto-upgraded to this incompatible version of Electron?

**Comment by rafeca** [comment]
> @rafeca: Since Electron 4 drops support for macOS 10.9, is there something we can do to ensure that Atom users on macOS 10.9 don't get auto-upgraded to this incompatible version of Electron?

I have no idea, maybe @ckerr , @jkleinsc or @BinaryMuse have some suggestions here, since other Electron apps that upgraded to v4 have probably experienced this same problem.

What we can do now is to start showing now banners on Atom running on OSX 10.9 informing users that we're going to drop support for this operating system very soon (maybe we can add this banner to v1.39 or v1.40).

**Comment by rafeca** [comment]
Regarding blocking 10.9 users from upgrading, this seems to be possible to do only by our `atom.io` backend system, which is the one that returns the potential list of versions to upgrade (This is based on [a Slack conversation with @jkleinsc](https://electronhq.slack.com/archives/CB759G125/p1561568279003900?thread_ts=1561565581.002300&cid=CB759G125)).

### client side changes

In order to do so, Atom needs to pass the OS version to `atom.io` when it asks if there are new versions. This is a very easy change which can be added [in this line](https://github.com/atom/atom/blob/master/src/main-process/auto-update-manager.js#L42-L44).

In order to make everything work, though, we need to do this change at least one the version before we ship Electron v4, so older clients that do not need to get updated are already sending the OS version.

My suggestion is to make these changes as soon as possible and cherry-pick them to the current beta to get them to stable soon.

### atom.io changes

Then, we need to change `atom.io` logic to take into account the OS version. This is going to be trickier to be able to also deal correctly with clients having much older versions of Atom.

In order to explain the logic a little bit clearer, I'm going to assume that the new parameter is added on Atom v1.40 and Electron v4 is added on Atom v1.42:

1. If the client does not pass the `os_version` param (old client), `atom.io` will return the latest v1.40.x version.
1. If the client passes `os_version=10.9` (and is a mac), `atom.io` will return the latest v1.41.x version.
1. If the client passes any other `os_version` param (or it's not a mac), `atom.io` will return the latest version (as it does now).

The same logic should be applied to each release channel (stable, beta and nightly).

Good thing is that if we build this logic in a configurable way, later it's going to be trivial to block e.g older Windows versions from upgrading if Electron drops support for them.

The `atom.io` changes can wait until we merge this branch (then we'll want to have them to prevent nightly users from upgrading).

@jasonrudolph thoughts?

**Comment by jasonrudolph** [comment]
@rafeca: That [approach](https://github.com/atom/atom/pull/19373#issuecomment-505969940) sounds good to me. 👍

**Comment by nathansobo** [comment]
This plan makes sense to me @rafeca.

One thing worth noting... Unless there's an API I don't know about, I think the best we'll be able to do is for determining the OS version is `require("os").release()`. That reports different version numbers than the 10.x used in Apple's marketing. Based on [this table of macOS releases](https://en.wikipedia.org/wiki/Darwin_%28operating_system%29#Release_history), it looks like 10.9 corresponds to a release number of `13.0.0` and 10.10 corresponds to `14.0.0`.

**Comment by rafeca** [comment]
While doing the regression testing I found an issue related to native tabs: when I enable them, doing operation with tabs becomes super slow and makes the editor unresponsive for ~5-10s. Same happens when doing operations with the native tabs (like merging different windows into a single window or splitting tabs into multiple windows).

I've created an issue to track it: https://github.com/atom/atom/issues/19611

**Comment by rafeca** [comment]
Ok, it looks like all the identified blockers are solved. In order to reduce the risk we're going to wait to merge this PR until just after the [Atom Railcar Release](https://github.com/atom/maintainers-log/issues/190) for 1.39 is done (which is expected on July 19th).

Until then, I'll be using a version of Atom built from this PR to see if I can find any other regression.

File: package.json
Comment (mfonville) [inline package.json]: 4.2.6 is also available

Comment (mfonville) [reply package.json]: and 4.2.7

**Review by IAMCRISTAL — APPROVED** [review]

File: package.json
Comment (mfonville) [reply package.json]: @rafeca I guess the 4.2.7 assets were not live yet when you tried to push the update

**Comment by mfonville** [comment]
Will Electron 4 be merged for the 1.41-dev branch?

File: package.json
Comment (rafeca) [reply package.json]: Yup, I tried 4.2.7 and it didn't work well (the release notes were not even published). I'm gonna upgrade to 4.2.7 now

**Comment by rafeca** [comment]
> Will Electron 4 be merged for the 1.41-dev branch?

Yup, I'm going to merge this PR once the CI passes. Since we've just rolled the railcars for v1.40.0, Electron v4 going to be available on v1.41 (probably from `v1.41.0-nightly1`).

**Comment by mfonville** [comment]
For 1.41 you can maybe upgrade Electron to the latest update 4.2.10?

**Comment by Chucky84** [comment]
[](url)

**Pull request merged**
